### ISSUE TITLE
Global objects not recognized if defined in a subsequent cell of a Jupyter notebook

### ISSUE THREAD

**Issue by altomani** [microsoft/pylance-release#4185]
## Environment data

-   Language Server version: 2023.3.40 (pyright 04471971)
-   OS and version: Windows 10
-   Python version (& distribution if applicable, e.g. Anaconda): Python 3.10.4

## Code Snippet

I have a notebook with the following cells:

Cell [1]

```python
class myclass:
    def __init__(self, a: int) -> None:
        self.a=a
        pass
```

Cell [2]

```python
def reset_myobj():
    myobj.a = 0
```

Cell [3]

```python
myobj = myclass(1)
reset_myobj()
```

## Expected behavior

No Pylance warnings

## Actual behavior

Pylance reports that

    "myobj" is not defined

in the body of `reset_myobj` in cell [2].
![image](https://user-images.githubusercontent.com/3511502/229483730-136d0bd3-0e38-465a-b48d-37f6cacbcc58.png)

The warning disappears in any of the following cases:

* `myobj` is defined in cell [1]
* `myobj` is defined in cell [2] before the definition of `reset_myobj`
* `myobj` is defined in cell [2] after the definition of `reset_myobj`

**Comment by heejaechang** [comment]
sounds like we need to put all global scope to the nested scope such as function

**Comment by heejaechang** [comment]
@erictraut 

if I have cells like

```python
# cell1
def foo():
    print(a)

# cell2
a = 1

# cell3
foo()

```

we show error on `print(a)` but running cells in sequence works fine and `cell3` print `1` as expected.

**Comment by erictraut** [comment]
Yeah, you're right. That won't result in a runtime exception.

It's still something I'd never do in a notebook because it affects the readability and unerstandability when a function refers to a symbol that hasn't been previously assigned a value.

**Comment by altomani** [comment]
I agree that relying on a not-yet-declared global is bad style, I reported this for two reasons:

1. if cell 2 and 3 are merged, then the warning disappears , and
2. I believe that it could be a relatively common pattern in interactive notebooks.

Regarding 1., the question is whether a notebook should be treated in the same way as a single file with all the cells concatenated linearly. That was my implicit assumption, but there are valid arguments to allow cells to depend only on previous cells.

Regarding point 2, in my case `myobj` is a singleton containing all the state and `reset_myobj` the callback associated to an ipywidgets button. The ipywidgets API makes it cumbersome to inject an argument in the callback, and it is common to modify global state in the callback.

**Comment by StellaHuang95** [comment]
This issue has been fixed in prerelease version 2026.2.100, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#20262100-3-April-2026-PreRelease)
